src/protocol/codecs/msgpack.rs

```diff
@@ -272,9 +273,84 @@ pub struct AgreementVote {
     pub sig: OneTimeSignature,
 }
 
+const CHECKSUM_LEN: usize = 4;
+const HASH_LEN: usize = 32;
+
+/// Public key address.
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct Address([u8; HASH_LEN]);
+
+impl Address {
+    /// Create a new [Address].
+    pub fn new(bytes: [u8; HASH_LEN]) -> Address {
+        Address(bytes)
+    }
+
+    /// Decode an address from a base64 string with a checksum.
+    pub fn from_string(string: &str) -> Result<Address, String> {
+        let checksum_address = match BASE32_NOPAD.decode(string.as_bytes()) {
+            Ok(decoded) => decoded,
+            Err(err) => return Err(format!("error decoding base32: {:?}", err)),
+        };
+
+        if checksum_address.len() != (HASH_LEN + CHECKSUM_LEN) {
+            return Err("input string is an invalid address. wrong length".to_string());
```

Comment:
It would be good (like eg. line 399) to have actual length of `checksum_address` (and maybe even expected) in error message.

Reply:
Done, thanks!